scylla/src/transport/load_balancing/default.rs

```diff
@@ -1,25 +1,334 @@
+        };
+
+        // If no token was available (or all the replicas for that token are down), try to pick
+        // some alive local node.
+        // If there was no preffered datacenter specified, all nodes are treated as local.
+        let nodes = self.preffered_node_set(cluster);
+        let picked = Self::pick_node(nodes, Self::is_alive);
+        if let Some(alive_local) = picked {
+            return alive_local;
+        }
+
+        // If a datacenter failover is possible, loosen restriction about locality.
+        if self.is_datacenter_failover_possible(&routing_info) {
+            let all_nodes = cluster.replica_locator().unique_nodes_in_global_ring();
+            let picked = Self::pick_node(all_nodes, Self::is_alive);
+            if let Some(alive_local) = picked {
+                return alive_local;
+            }
+        }
+
+        // Previous checks imply that every node we could have selected is down.
+        // Let's try to return a down node that wasn't disabled.
+        let picked = Self::pick_node(nodes, |node| node.is_enabled());
+        if let Some(down_but_enabled_node) = picked {
+            return down_but_enabled_node;
+        }
+
+        // Every node is disabled. This could be due to a bad host filter - configuration error.
+        //
+        // There is at least one node in cluster (this is a `ClusterData` invariant), and also at
+        // least one node in each datacenter (because the `Node` is a source of all datacenter
+        // informations), so the unwrap is safe here.
+        nodes.first().unwrap()
+    }
+
+    fn fallback<'a>(&'a self, query: &'a QueryInfo, cluster: &'a ClusterData) -> FallbackPlan<'a> {
+        let routing_info = self.routing_info(query, cluster);
+
+        // If token is available, get a shuffled list of alive replicas.
+        let maybe_replicas = if let Some(ts) = &routing_info.token_with_strategy {
+            let local_replicas = self.shuffled_replicas(ts, true, Self::is_alive, cluster);
+
+            // If a datacenter failover is possible, loosen restriction about locality.
+            let maybe_remote_replicas = if self.is_datacenter_failover_possible(&routing_info) {
+                let remote_replicas = self.shuffled_replicas(ts, false, Self::is_alive, cluster);
+                Either::Left(remote_replicas)
+            } else {
+                Either::Right(std::iter::empty())
+            };
+
+            // Produce an iterator, prioritizes local replicas.
+            // If preferred datacenter is not specified, every replica is treated as a local one.
+            Either::Left(local_replicas.chain(maybe_remote_replicas))
+        } else {
+            Either::Right(std::iter::empty::<NodeRef<'a>>())
+        };
+
+        // Get a list of all local alive nodes, and apply a round robin to it
+        let local_nodes = self.preffered_node_set(cluster);
+        let robined_local_nodes = Self::round_robin_nodes(local_nodes, Self::is_alive);
```

Comment:
Round robin has the problem of overloads in case of node failure.
If the usual round robin order is A->B->C and A fails, B will take over all of A's requests. It would be better to try nodes in random order, then in case of A's failure the load that used to be handled by A will be shared equally among B and C.

Reply:
As we have no quick idea how to mitigate this problem yet, let's put off this problem until a follow-up.